### PR TITLE
Add health check endpoint

### LINKED ISSUE #8017: Provide a http healthcheck endpoint that doesn't log at INFO
This will reduce unneeded logging in some cloud environments, as we don't need to log on a ~2s basis that we are available.

General request logging should be unaffected by this.

### PR DESCRIPTION
**Pull request by erikjohnston**
This gets added under `/health` on *every* HTTP listener. Requests to `/health` get logged at DEBUG rather than INFO.

Question: Where is the best place to document this?

Fixes #8017

### CODE REVIEW

File: tests/rest/test_health.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+from synapse.rest.health import HealthResource
+
+from tests import unittest
+
+
+class HealthCheckTests(unittest.HomeserverTestCase):
+    def setUp(self):
+        super().setUp()
+
+        # replace the JsonResource with a WellKnownResource
```
Comment (clokep) [inline tests/rest/test_health.py:26]:
```suggestion
        # replace the JsonResource with a HealthResource.
```

**Comment by clokep** [comment]
> This gets added under `/health` on _every_ HTTP endpoint.

Not 100% sure what this means. I think you mean for each enabled `resource`?

**Comment by clokep** [comment]
> Question: Where is the best place to document this?

Probably near the reverse proxy information?

**Comment by erikjohnston** [comment]
> > This gets added under `/health` on _every_ HTTP endpoint.
> 
> Not 100% sure what this means. I think you mean for each enabled `resource`?

Woops, yes I mean on every configured HTTP listener

**Comment by clokep** [comment]
> Requests to `/health` get logged at INFO rather than DEBUG.

The code here seems to say the opposite:

```python
log_level = logging.INFO if (self.path != b"/health") else logging.DEBUG
```

I think the code is right and the PR description is wrong (based on the documentation added)?

**Comment by erikjohnston** [comment]
> > Requests to `/health` get logged at INFO rather than DEBUG.
> 
> The code here seems to say the opposite:
> 
> ```python
> log_level = logging.INFO if (self.path != b"/health") else logging.DEBUG
> ```
> 
> I think the code is right and the PR description is wrong (based on the documentation added)?

This is just embarrassing now. Yes, the description is wrong.

File: docs/reverse_proxy.md [hunk 2]
```diff
@@ -139,3 +139,10 @@ client IP addresses are recorded correctly.
 Having done so, you can then use `https://matrix.example.com` (instead
 of `https://matrix.example.com:8448`) as the "Custom server" when
 connecting to Synapse from a client.
+
+
+## Health check endpoint
+
+Synapse exposes a `/health` endpoint on every configured HTTP listener that
+always returns 200 OK (and doesn't get logged) for use by reverse proxies as a
+health check endpoint.
```
Comment (clokep) [inline docs/reverse_proxy.md:148]: I find this a bit confusing, the following might be clearer? Not sure it is an improvement.

```suggestion
Synapse exposes a health check endpoint for use by reverse proxies.
Each configured HTTP listener has a `/health` endpoint which always returns
200 OK (and doesn't get logged).
```

Maybe I'm just being a bit thick, but this means that if I have Synapse running on localhost:8080 I would be able to do the health check at: http://localhost:8080/health, correct? It isn't http://localhost:8080/_matrix/client/r0/health?

Comment (erikjohnston) [reply docs/reverse_proxy.md]: Yup, its http://localhost:8080/health

**Comment by michaelkaye** [comment]
Wondering if we should add this to a HEALTHCHECK line in the docker files.

I have mixed feelings about re-invoking curl each time the healthcheck fires to ensure the container is happy, but it seems to be the way to do it.

**Comment by richvdh** [comment]
related: https://github.com/matrix-org/synapse/pull/7995

**Comment by clokep** [comment]
> Wondering if we should add this to a HEALTHCHECK line in the docker files.
> 
> I have mixed feelings about re-invoking curl each time the healthcheck fires to ensure the container is happy, but it seems to be the way to do it.

Looks like #8147 implements this.

**Pull request merged**